runtime/common/src/custom_transactor.rs

```diff
@@ -0,0 +1,158 @@
+
+use sp_runtime::codec::FullCodec;
+use sp_runtime::{
+	traits::{Convert, MaybeSerializeDeserialize, SaturatedConversion},
+};
+use sp_std::{
+	cmp::{Eq, PartialEq},
+	fmt::Debug,
+	marker::PhantomData,
+	prelude::*,
+	result,
+};
+
+use xcm::v3::{prelude::*, Error as XcmError, MultiAsset, MultiLocation, Result};
+use xcm_executor::{
+	traits::{Convert as MoreConvert, MatchesFungible, TransactAsset},
+	Assets,
+};
+use orml_xcm_support::{UnknownAsset as UnknownAssetT,OnDepositFail};
+
+
+/// Asset transaction errors.
+enum Error {
+	/// Failed to match fungible.
+	FailedToMatchFungible,
+	/// `MultiLocation` to `AccountId` Conversion failed.
+	AccountIdConversionFailed,
+	/// `CurrencyId` conversion failed.
+	CurrencyIdConversionFailed,
+}
+
+impl From<Error> for XcmError {
+	fn from(e: Error) -> Self {
+		match e {
+			Error::FailedToMatchFungible => XcmError::FailedToTransactAsset("FailedToMatchFungible"),
+			Error::AccountIdConversionFailed => XcmError::FailedToTransactAsset("AccountIdConversionFailed"),
+			Error::CurrencyIdConversionFailed => XcmError::FailedToTransactAsset("CurrencyIdConversionFailed"),
+		}
+	}
+}
+
+pub trait AutomationPalletConfig {
+	fn matches_asset(asset: &MultiAsset) -> Option<u128>;
+	fn matches_beneficiary(beneficiary_location: &MultiLocation) -> Option<(u8, [u8;32])>;
+	fn callback(length: u8, data: [u8;32], amount: u128) -> Result;
+}
+
+#[allow(clippy::type_complexity)]
+pub struct CustomMultiCurrencyAdapter<
```

Comment:
This draws heavy inspiration from ORMLs [MultiCurrencyAdapter](https://github.com/open-web3-stack/open-runtime-module-library/blob/84c58de7820d49acf81e7e4b6a5e28f8572bef95/xcm-support/src/currency_adapter.rs#L105) which is good because there is obviously no need to reinvent the wheel here. I'm just wondering if it wouldn't be better to change this struct so that it takes two traits, 1) the existing `MultiCurrencyAdapter` and the `AutomationPalletConfigT`. And then we make the `deposit/withdraw/transfer_asset()` functions just call the methods on the `MultiCurrencyAdapter` that was passed to us but for the `deposit_asset()` we add the extra automation pallet check before calling the 'parent' function.
Would this work, what do you think @gianfra-t? This way we can deduplicate some code and profit from upstream changes made to the `MultiCurrencyAdapter` of orml, if any.

Reply:
That is a very good point, I can't believe I didn't think on that. We can just require that trait 1) just implements `deposit` and `withdraw` so we can later switch it for something else.

Reply:
I changed it so the "WrappedTransactor" (`MultiCurrencyAdapter` in this case) has to implement xcm's `TransactAsset`, which is anyways what `AssetTransactor` must implement so I guess this is as generic as we can go.